src/Razor/test/Microsoft.VisualStudio.LanguageServices.Razor.Test/Cohost/CohostDocumentCompletionEndpointTest.cs

```diff
@@ -0,0 +1,486 @@
+
+                <div>@DateTime.$$</div>
+
+                The end.
+                """,
+             completionContext: new RoslynVSInternalCompletionContext()
+             {
+                 InvokeKind = RoslynVSInternalCompletionInvokeKind.Typing,
+                 TriggerCharacter = ".",
+                 TriggerKind = RoslynCompletionTriggerKind.TriggerCharacter
+             },
+             expectedItemLabels: ["DaysInMonth", "IsLeapYear", "Now"]);
+    }
+
+    [Fact]
+    public async Task CSharpClassesInCodeBlock()
+    {
+        await VerifyCompletionListAsync(
+            input: """
+                This is a Razor document.
+
+                <div></div>
+
+                @code{ $$ }
+
+                The end.
+                """,
+             completionContext: new RoslynVSInternalCompletionContext()
+             {
+                 InvokeKind = RoslynVSInternalCompletionInvokeKind.Explicit,
+                 TriggerCharacter = null,
+                 TriggerKind = RoslynCompletionTriggerKind.Invoked
+             },
+             expectedItemLabels: ["char", "DateTime", "Exception"]);
+    }
+
+    [Fact]
+    public async Task CSharpClassMembersInCodeBlock()
+    {
+        await VerifyCompletionListAsync(
+            input: """
+                This is a Razor document.
+
+                <div></div>
+
+                @code{
+                    void foo()
+                    {
+                        DateTime.$$
+                    }
+                }
+
+                The end.
+                """,
+             completionContext: new RoslynVSInternalCompletionContext()
+             {
+                 InvokeKind = RoslynVSInternalCompletionInvokeKind.Typing,
+                 TriggerCharacter = ".",
+                 TriggerKind = RoslynCompletionTriggerKind.TriggerCharacter
+             },
+             expectedItemLabels: ["DaysInMonth", "IsLeapYear", "Now"]);
+    }
+
+    // Tests MarkupTransitionCompletionItemProvider
+    [Fact]
+    public async Task CSharpMarkupTransitionAndTagHelpersInCodeBlock()
+    {
+        await VerifyCompletionListAsync(
+            input: """
+                This is a Razor document.
+
+                <div></div>
+
+                @code{
+                    void foo()
+                    {
+                        <$$
+                    }
+                }
+
+                The end.
+                """,
+             completionContext: new RoslynVSInternalCompletionContext()
+             {
+                 InvokeKind = RoslynVSInternalCompletionInvokeKind.Typing,
+                 TriggerCharacter = "<",
+                 TriggerKind = RoslynCompletionTriggerKind.TriggerCharacter
+             },
+             expectedItemLabels: ["text", "EditForm", "InputDate"]);
+    }
+
+    [Fact]
+    public async Task RazorDirectives()
+    {
+        var expectedDirectiveLabels =
+            DirectiveCompletionItemProvider.s_singleLineDirectiveSnippets.Keys
```

Comment:
I am very much not a fan of using real product code for expected test outputs. It means the test is validating that we pass data from one place to another, rather than validating that we offer the user the things we expect to offer. If a bad merge means we accidentally remove something from `DirectiveCompletionItemProvider`, the unit test will not help us.

By my count there will be 10 things in `expectedDirectiveLabels`, so I think it's completely reasonable to just hard code the list of 10 items here, and makes for a much better test.

Reply:
My thinking was that losing a directive in merge would be as likely as forgetting to add it in a test. Also even if we have a hardcoded list here, we still could lose one of them in the merge. At the same time I don't feel strongly one way or the other here, and since Dustin feels the same way as you David, I will go with your suggestion.